File: src/main/java/io/github/eone666/telegramnotifier/features/PlayersList.java
package io.github.eone666.telegramnotifier.features;

import io.github.eone666.telegramnotifier.TelegramNotifier;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.json.simple.JSONObject;
import java.util.Collection;
import java.util.HashSet;
import java.util.stream.Collectors;

public class PlayersList {

    private TelegramNotifier plugin = null;
    private Collection<Player> players = new HashSet<Player>();
    String text = null;

    public PlayersList (TelegramNotifier pluginInstance) {
        plugin = pluginInstance;
    }

    private void buildText() {
        int playersCount = players.size();
        String playersTextByCount = playersCount == 0 ? "No players online" : "Players online:";
        String playerNames = players.stream().map(player -> player.getName()).map(Object::toString).collect(Collectors.joining("\n"));
        text = playersTextByCount + "\n" + playerNames;
        if(plugin.config.getIsPlayersListHeaderEnabled()){
            text = plugin.config.getPlayersListHeaderText() + "\n" + text;
        }
    }

    private void sendNewMessageAndPin () {
        JSONObject response = plugin.tg.sendMessage(plugin.config.getChatId(), text, "markdown");

        boolean isOk = Boolean.valueOf(response.get("ok").toString());
        if(isOk){
            JSONObject resultObject = (JSONObject) response.get("result");
            int messageId = Integer.parseInt(resultObject.get("message_id").toString());
            plugin.config.setPlayersListMessageId(messageId);
            plugin.tg.pinChatMessage(plugin.config.getChatId(), messageId);
        }
    }

    private void editMessage () {
        JSONObject response = plugin.tg.editMessageText(
                plugin.config.getChatId(),
                plugin.config.getPlayersListMessageId(),
                text, "markdown"
        );

        boolean isOk = Boolean.valueOf(response.get("ok").toString());
        if(!isOk){
            int errorCode = Integer.valueOf(response.get("error_code").toString());
            String description = response.get("description").toString();

            if (errorCode == 400 && description == "Bad Request: message to edit not found") {
                Bukkit.getLogger().info("Sending new message");
                sendNewMessageAndPin();
            }
        }
    }

    private void updateMessage () {
        buildText();
        if(plugin.config.getPlayersListMessageId() == 0) {
            sendNewMessageAndPin();
        } else {
            editMessage();
        }}

    public void join (Player player) {
        players.add(player);
        updateMessage();
    }

    public void quit (Player player) {
        players.remove(player);
        updateMessage();
    }

    public void init () {
        updateMessage();
    }

    public void disable () {
        players.clear();
        updateMessage();
    }

}

File: src/main/kotlin/io/github/eone666/telegramnotifier/features/PlayersList.kt
package io.github.eone666.telegramnotifier.features

import com.elbekd.bot.model.toChatId
import io.github.eone666.telegramnotifier.pluginInstance
import org.bukkit.entity.Player
import java.util.stream.Collectors
import com.elbekd.bot.types.ParseMode.Markdown
import com.github.shynixn.mccoroutine.bukkit.launch

class PlayersList {
    private val config = pluginInstance.config
    private val tg = pluginInstance.bot?.tg

    private val players: MutableCollection<Player> = HashSet()
    private var text: String = ""
    private fun buildText() {
        val playersCount = players.size
        val playerNames = players.stream().map { obj: Player -> obj.name }.collect(Collectors.joining("\n"))

        text = if (playersCount == 0) "No players online" else "Players online:\n$playerNames"

        if (config.isPlayersListHeaderEnabled.get()) {
            text = "${config.playersListHeaderText.get()}\n$text"
        }

        if (config.isPlayersListFooterEnabled.get()) {
            text = "$text\n\n${config.playersListFooterText.get()}"
        }
    }

    private suspend fun sendNewMessageAndPin() {
        try {
            val response = tg?.sendMessage(
                chatId = config.chatId.get().toChatId(),
                disableNotification = config.isNotificationsSilentModeEnabled.get(),
                text = text,
                parseMode = Markdown,
                disableWebPagePreview = true
            )
            config.playersListMessageId.set(response!!.messageId)
            config.save()
            pluginInstance.logger.info("New message sent")
        } catch (err: Throwable) {
            pluginInstance.logger.warning(err.message)
        }

        try {
            tg?.pinChatMessage(
                chatId = config.chatId.get().toChatId(),
                disableNotification = config.isNotificationsSilentModeEnabled.get(),
                messageId = config.playersListMessageId.get()
            )
            pluginInstance.logger.info("Message pinned")
        } catch (err: Throwable) {
            pluginInstance.logger.warning(err.message)
        }
    }

    private suspend fun editMessage() {
        try {
            tg?.editMessageText(
                chatId = config.chatId.get().toChatId(),
                messageId = config.playersListMessageId.get(),
                text = text,
                parseMode = Markdown,
                disableWebPagePreview = true
            )
        } catch (err: Throwable) {
            pluginInstance.logger.warning(err.message)
        }

    }

    private suspend fun updateMessage() {
        buildText()
        if (config.playersListMessageId.get() == 0.toLong()) {
            sendNewMessageAndPin()
        } else {
            editMessage()
        }
    }

    suspend fun add(player: Player) {
        if (config.isPlayersListEnabled.get()) {
            players.add(player)
            updateMessage()
        }
    }

    suspend fun remove(player: Player) {
        if (config.isPlayersListEnabled.get()) {
            players.remove(player)
            updateMessage()
        }
    }

    fun update() {
        if (config.isPlayersListEnabled.get()) {
            pluginInstance.launch {
                updateMessage()
            }
        }
    }

    fun clear() {
        if (config.isPlayersListEnabled.get()) {
            players.clear()
            pluginInstance.launch {
                updateMessage()
            }
        }
    }
}
